File: sections/73_convex_geometry_minkowski/problems/d_convex_polygon_distance/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;
using int128 = __int128_t;

struct Point {
    int64 x, y;
    Point operator+(Point other) const { return {x + other.x, y + other.y}; }
    Point operator-(Point other) const { return {x - other.x, y - other.y}; }
};

int128 cross(Point a, Point b) {
    return int128(a.x) * b.y - int128(a.y) * b.x;
}

vector<Point> normalize(vector<Point> polygon) {
    int start = min_element(polygon.begin(), polygon.end(), [](Point a, Point b) {
        return pair(a.y, a.x) < pair(b.y, b.x);
    }) - polygon.begin();
    rotate(polygon.begin(), polygon.begin() + start, polygon.end());
    return polygon;
}

vector<Point> minkowski(vector<Point> a, vector<Point> b) {
    a = normalize(move(a));
    b = normalize(move(b));
    int n = a.size(), m = b.size();
    vector<Point> edge_a(n), edge_b(m);
    for (int i = 0; i < n; ++i) edge_a[i] = a[(i + 1) % n] - a[i];
    for (int i = 0; i < m; ++i) edge_b[i] = b[(i + 1) % m] - b[i];

    vector<Point> result = {a[0] + b[0]};
    int i = 0, j = 0;
    while (i < n || j < m) {
        Point step;
        if (j == m || (i < n && cross(edge_a[i], edge_b[j]) > 0)) {
            step = edge_a[i++];
        } else if (i == n || cross(edge_a[i], edge_b[j]) < 0) {
            step = edge_b[j++];
        } else {
            step = edge_a[i++] + edge_b[j++];
        }
        result.push_back(result.back() + step);
    }
    result.pop_back();
    return result;
}

long double segment_distance(Point a, Point b) {
    long double dx = b.x - a.x;
    long double dy = b.y - a.y;
    long double denominator = dx * dx + dy * dy;
    long double parameter = -(a.x * dx + a.y * dy) / denominator;
    parameter = max((long double)0, min((long double)1, parameter));
    long double x = a.x + parameter * dx;
    long double y = a.y + parameter * dy;
    return sqrt(x * x + y * y);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n;
    vector<Point> a(n);
    for (auto& point : a) cin >> point.x >> point.y;
    cin >> m;
    vector<Point> negative_b(m);
    for (auto& point : negative_b) {
        cin >> point.x >> point.y;
        point.x = -point.x;
        point.y = -point.y;
    }

    vector<Point> difference = minkowski(move(a), move(negative_b));
    bool contains_origin = true;
    for (int i = 0; i < (int)difference.size(); ++i) {
        Point edge = difference[(i + 1) % difference.size()] - difference[i];
        Point toward_origin{-difference[i].x, -difference[i].y};
        if (cross(edge, toward_origin) < 0) contains_origin = false;
    }

    long double answer = 0;
    if (!contains_origin) {
        answer = numeric_limits<long double>::infinity();
        for (int i = 0; i < (int)difference.size(); ++i) {
            answer = min(answer, segment_distance(
                difference[i], difference[(i + 1) % difference.size()]));
        }
    }
    cout << fixed << setprecision(12) << answer << '\n';
}

File: sections/73_convex_geometry_minkowski/problems/d_convex_polygon_distance/solution.py
import math
import sys


def cross(a, b):
    return a[0] * b[1] - a[1] * b[0]


def subtract(a, b):
    return a[0] - b[0], a[1] - b[1]


def normalize(polygon):
    start = min(range(len(polygon)), key=lambda i: (polygon[i][1], polygon[i][0]))
    return polygon[start:] + polygon[:start]


def minkowski(a, b):
    a = normalize(a)
    b = normalize(b)
    edge_a = [subtract(a[(i + 1) % len(a)], a[i]) for i in range(len(a))]
    edge_b = [subtract(b[(i + 1) % len(b)], b[i]) for i in range(len(b))]
    result = [(a[0][0] + b[0][0], a[0][1] + b[0][1])]
    i = j = 0
    while i < len(edge_a) or j < len(edge_b):
        if j == len(edge_b) or (
            i < len(edge_a) and cross(edge_a[i], edge_b[j]) > 0
        ):
            step = edge_a[i]
            i += 1
        elif i == len(edge_a) or cross(edge_a[i], edge_b[j]) < 0:
            step = edge_b[j]
            j += 1
        else:
            step = (edge_a[i][0] + edge_b[j][0], edge_a[i][1] + edge_b[j][1])
            i += 1
            j += 1
        result.append((result[-1][0] + step[0], result[-1][1] + step[1]))
    result.pop()
    return result


def segment_distance(a, b):
    dx, dy = b[0] - a[0], b[1] - a[1]
    parameter = -(a[0] * dx + a[1] * dy) / (dx * dx + dy * dy)
    parameter = min(1.0, max(0.0, parameter))
    x = a[0] + parameter * dx
    y = a[1] + parameter * dy
    return math.hypot(x, y)


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    at = 0
    n = data[at]
    at += 1
    a = [(data[at + 2 * i], data[at + 2 * i + 1]) for i in range(n)]
    at += 2 * n
    m = data[at]
    at += 1
    negative_b = [(-data[at + 2 * i], -data[at + 2 * i + 1]) for i in range(m)]
    difference = minkowski(a, negative_b)
    contains_origin = all(
        cross(
            subtract(difference[(i + 1) % len(difference)], difference[i]),
            (-difference[i][0], -difference[i][1]),
        ) >= 0
        for i in range(len(difference))
    )
    answer = 0.0 if contains_origin else min(
        segment_distance(difference[i], difference[(i + 1) % len(difference)])
        for i in range(len(difference))
    )
    print(f"{answer:.12f}")


if __name__ == "__main__":
    main()
